### ISSUE TITLE
Wrong image scaling in SwiftUI (layout size is calculated in pixels)

### ISSUE THREAD

**Issue by Kinark** [kean/Nuke#746]
As of now LazyImage renders remote images with the scale of 1 instead of using the scale of the screen. From what I could tell, there's no way of showing the image with the proper scaling without using `.resizable()` modifier which degrades performance as you and a few other devs talked about [here](https://mastodon.social/@christianselig/109881390243372020). I've tried:

* Using thumbnail options in userInfo
* Setting the screen scale manually in userInfo
* Using a `Resize` processor

...but nothing works except if `.resizable()` is present.

`AsyncImage` has an optional `scale` parameter that fixes that, but there's no similar alternative in Nuke.

Take [this](https://i.imgur.com/k8gZbJ4.png) image as a sample, which has a size of 96x96. In my iPhone 14 Pro, which has a screen scale of 3, it should show as a rectangle with the size of 32x pts, but instead it renders as a 96x pts image. The same happens with AsyncImage if `scale` param is not set.

Here's a code sample which tests all the alternatives:

```
let req = ImageRequest(url: URL(string: "https://i.imgur.com/k8gZbJ4.png"), userInfo: [.thumbnailKey: ImageRequest.ThumbnailOptions(size: .init(width: 32, height: 32), unit: .points, contentMode: .aspectFill), .scaleKey: 3])
VStack {
  LazyImage(request: req)
  LazyImage(request: req)
    .processors([.resize(width: 32)])
  LazyImage(url:  URL(string: "https://i.imgur.com/k8gZbJ4.png")!)
    .processors([.resize(width: 32)])
  LazyImage(url:  URL(string: "https://i.imgur.com/k8gZbJ4.png")!)
  AsyncImage(url: URL(string: "https://i.imgur.com/k8gZbJ4.png")!, scale: 3)
  AsyncImage(url: URL(string: "https://i.imgur.com/k8gZbJ4.png")!)
}
```

This is the the result:
![Screenshot 2024-01-09 at 23 45 51](https://github.com/kean/Nuke/assets/17537081/f4c10312-1d6b-47b2-874a-1a7a4cd9139a)

Is there anything I'm missing or is the expected behavior?

**Comment by kean** [comment]
I debugged the following scenario:

```swift
let request = ImageRequest(url: URL(string: "https://i.imgur.com/k8gZbJ4.png"), userInfo: [..scaleKey: 3])
LazyImage(request: request)
```

The image displayed by the `LazyImage` appears to have the correct size and scale.

(lldb) po imageContainer?.image
▿ Optional<UIImage>
  - some : <UIImage:0x60000302c120 anonymous {32, 32} renderingMode=automatic(original)>
(lldb) po imageContainer?.image.scale
▿ Optional<CGFloat>
  - some : 3.0

When I run it in a simulator, the image has the correct size:

<img width="1112" alt="Screenshot 2024-01-09 at 10 06 37 PM" src="https://github.com/kean/Nuke/assets/1567433/a590a6da-1f57-4b1f-be7f-5b8e9ded3c02">

**Comment by kean** [comment]
I think the issue it due to coalescing and/or caching of the requests. The framework tries to minimize the amount of work and reuse the existing cached images and download and processing tasks. The issue is that the `.scaleKey` isn't used in the cache or coalescing keys. I don't think it was ever designed for a scenario where someone would request the same image with different scales.

I haven't fully confirmed this hypothesis as it's a bit tricky to test, but it seems to be the case.

> Setting the screen scale manually in userInfo

Yes, that's the intended way to use it, and is a replacement for the `scale` parameter in `AsyncImage`. Unlike `AsyncImage`, Nuke provides a ton of different options, so, of course, it made sense to extend the existing options rather than adding an ad-hoc parameter to the `LazyImage` type itself.

**Comment by Kinark** [comment]
> I debugged the following scenario:
> 
> ```swift
> let request = ImageRequest(url: URL(string: "https://i.imgur.com/k8gZbJ4.png"), userInfo: [..scaleKey: 3])
> LazyImage(request: request)
> ```
> 
> The image displayed by the `LazyImage` appears to have the correct size and scale.
> 
> (lldb) po imageContainer?.image ▿ Optional
> 
> * some : <UIImage:0x60000302c120 anonymous {32, 32} renderingMode=automatic(original)>
>   (lldb) po imageContainer?.image.scale
>   ▿ Optional
> * some : 3.0
> 
> When I run it in a simulator, the image has the correct size:
> 
> <img alt="Screenshot 2024-01-09 at 10 06 37 PM" width="1112" src="https://private-user-images.githubusercontent.com/1567433/295427669-a590a6da-1f57-4b1f-be7f-5b8e9ded3c02.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgyNDYwMjUsIm5iZiI6MTcwODI0NTcyNSwicGF0aCI6Ii8xNTY3NDMzLzI5NTQyNzY2OS1hNTkwYTZkYS0xZjU3LTRiMWYtYmU3Zi01YjhlOWRlZDNjMDIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMThUMDg0MjA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2M4ODY0YWQxNzM3ZDQ3Yzc5NTg1NzdkYTE2YjZmYjkwNTc1MWU5MDJiM2M0OGIzNDQyNDNiZDc3YzFlMzNmOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.CW1PKMJMeahHLt5B3i4vDBcaBMY-stAzulOznUguxpc">

Sorry for the late response!

So I guess that the problem happens when we need to rescale the image with either `thumbnailKey` or `resize` preprocessor. Like, the image I provided has a size of 96x96px, which displays nicely in a case where we need the image to be 32x32pt, but what if my image has a size os 256x256px and I need to display a thumbnail of it in the size of 32x32pts? How can I preprocess the image but still keep the correct aspect ratio?

**Comment by kean** [comment]
My suggestions would be to either make sure you use the same scale every time OR disable coalescing using pipeline's configuration. 

I'm also going to mark it as a defect because the scale parameter should never break caching or coalescing.

There may also be a need for a different API or implementation for setting the scale. Maybe it should be something similar to the `AsyncImage` API. It should _probably_ be safe and quick to fetch the image and then change the scale right before the display every time the image is displayed.

**Comment by kean** [comment]
The coalescing issue were fixed in [Nuke 12.6](https://github.com/kean/Nuke/releases/tag/12.6.0).
